src/TestFramework/MSTest.Core/Attributes/EnumDataSourceAttribute.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestTools.UnitTesting
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+
+    /// <summary>
+    /// Attribute to define dynamic data for a test method.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+    public class EnumDataSourceAttribute : Attribute, ITestDataSource
+    {
+        private readonly Type enumDataSource;
+        private readonly object[] exclusions;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EnumDataSourceAttribute"/> class.
+        /// </summary>
+        /// <param name="enumDataSource">
+        /// The type of the <see cref="Enum"/> to iterate for the test.
+        /// </param>
+        public EnumDataSourceAttribute(Type enumDataSource)
+            : this(enumDataSource, new object[0])
+        {
+            // Need to have this constructor explicitely to fix a CLS compliant error.
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EnumDataSourceAttribute"/> class.
+        /// </summary>
+        /// <param name="enumDataSource">
+        /// The type of the <see cref="Enum"/> to iterate for the test.
+        /// </param>
+        /// <param name="exclusions">
+        /// Values to exclude during the iteration.
+        /// </param>
+        public EnumDataSourceAttribute(Type enumDataSource, params object[] exclusions)
```

Comment:
I would have loved to create a generic attribute `EnumDataSourceAttribute<T> where T: System.Enum`, and then enforce the type here (`params T[] exclusions`), but sadly en attribute cannot be a generic type :(